File: C++/Level.3/베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

bool cmp(pair<int, string> a, pair<int, string> b)
{
    return a.first > b.first;
}

bool compare(pair<int, int> a, pair<int, int> b)
{
    return a.first > b.first;
}
vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    map<string, int> song_table;
    int song_number = genres.size();
    
    for(int i = 0; i < song_number; ++i)
    {
        song_table[genres[i]] += plays[i];
    }
    
    map<string, int>::iterator iter;
    string key = "";
    int played = 0;
    vector<pair<int, string>> vec;
    
    for(iter = song_table.begin(); iter != song_table.end(); ++iter)
    {
        vec.push_back({iter->second, iter->first});
    }
    sort(vec.begin(), vec.end(), cmp);
    
    for(int i = 0; i < vec.size(); ++i)
    {
        string key = vec[i].second;
        vector<pair<int, int>> temp;
        for(int j = 0; j < song_number; ++j)
        {
            if(genres[j] == key)
            {
                temp.push_back({plays[j], j});
            }
        }
        sort(temp.begin(), temp.end(), compare);
        if(temp.size() == 1)
        {
            answer.push_back(temp[0].second);
        }
        else
        {
            answer.push_back(temp[0].second);
            answer.push_back(temp[1].second);
        }
    }

    return answer;
}

File: JavaScript/Level.3/베스트앨범.js
function solution(genres, plays) {
    const len = genres.length;
    let playedNum = new Map();
    let list = new Map();
    for (let i = 0; i < len; ++i) {
        playedNum.has(genres[i])
            ? playedNum.set(genres[i], playedNum.get(genres[i]) + plays[i])
            : playedNum.set(genres[i], plays[i]);
        list.has(genres[i])
            ? list.set(genres[i], [...list.get(genres[i]), [plays[i], i]])
            : list.set(genres[i], [[plays[i], i]]);
    }
    let map2Array = Array.from(playedNum);
    map2Array.sort((a, b) => {
        return b[1] - a[1];
    });
    for (let [key, value] of list) {
        value.sort((a, b) => {
            if (a[0] === b[0]) return a[1] - b[1];
            return b[0] - a[0];
        });
    }
    let answer = [];
    for (let key of map2Array) {
        if (list.get(key[0]).length === 1) {
            answer.push(list.get(key[0])[0][1]);
        } else {
            answer.push(list.get(key[0])[0][1]);
            answer.push(list.get(key[0])[1][1]);
        }
    }

    return answer;
}
